### PR TITLE
Integration of HQS automatic backend selection

### PR DESCRIPTION
**Pull request by David-Kreplin**

### CODE REVIEW

**Comment by MoritzWillmann** [comment]
Is this something we want in sQUlearn rather than autoqml-lib?

**Comment by David-Kreplin** [comment]
Currently update the branch to current squlearn level

File: src/squlearn/util/executor.py [hunk 1]
```diff
@@ -1475,6 +1681,116 @@ def set_seed_for_primitive(self, seed: int = 0):
 
         self._set_seed_for_primitive = seed
 
+    def select_backend(self, circuit, **options):
+        from ..encoding_circuit.encoding_circuit_base import (
+            EncodingCircuitBase,
+        )  # check why not outside
+        from ..encoding_circuit.transpiled_encoding_circuit import TranspiledEncodingCircuit
```
Comment (MoritzWillmann) [inline src/squlearn/util/executor.py:1688]: move to top of file

File: src/squlearn/util/executor.py [hunk 2]
```diff
@@ -1475,6 +1681,116 @@ def set_seed_for_primitive(self, seed: int = 0):
...
+        min_num_qubits = options.get("min_num_qubits", None)
+        max_num_qubits = options.get("max_num_qubits", None)
+        cost_function = options.get("cost_function", None)
+        optimization_level = options.get("optimization_level", 3)
+        n_trials_transpile = options.get("n_trials_transpile", 1)
+        call_limit = options.get("call_limit", int(3e7))
+        verbose = options.get("verbose", False)
+        logger = self._logger
+
+        AutoSelBack = AutoSelectionBackend(
```
Comment (MoritzWillmann) [inline src/squlearn/util/executor.py:1701]:
```suggestion
        auto_selection_backend= AutoSelectionBackend(
```

File: src/squlearn/util/execution/backend_auto_selection.py [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+                optimization_level=self.optimization_level,
+            )
+
+        if final_circuit.layout.final_layout:
+            self._print("Last transpilation: sub-layout has been changed")
+            return (final_circuit.layout.final_layout, score), final_circuit, least_busy_backend
+        else:
+            return (final_circuit.layout.initial_layout, score), final_circuit, least_busy_backend
+
+    def evaluate_via_HQAA(
```
Comment (MoritzWillmann) [inline src/squlearn/util/execution/backend_auto_selection.py:407]:
```suggestion
    def _evaluate_via_HQAA(
```

File: src/squlearn/util/execution/backend_auto_selection.py [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+            if backend_qubits >= circuit.num_qubits:
+                # Add one backend per family of processor.
+                backend_by_family[backend_processor_family] = backend
+
+        if len(backend_by_family) == 0:
+            raise NoSuitableBackendError("No suitable backend found with given parameters")
+        else:
+            return backend_by_family
+
+    def tensor_circuits(self, circuit_list: List[QuantumCircuit]) -> QuantumCircuit:
```
Comment (MoritzWillmann) [inline src/squlearn/util/execution/backend_auto_selection.py:248]:
```suggestion
    def _tensor_circuits(self, circuit_list: List[QuantumCircuit]) -> QuantumCircuit:
```

File: src/squlearn/util/executor.py [hunk 5]
```diff
@@ -1475,6 +1681,116 @@ def set_seed_for_primitive(self, seed: int = 0):
...
+            cost_function=cost_function,
+            optimization_level=optimization_level,
+            n_trials_transpile=n_trials_transpile,
+            call_limit=call_limit,
+            verbose=verbose,
+            logger=logger,
+        )
+
+        mode = options.get("mode", self._auto_backend_mode)
+        useHQAA = options.get("useHQAA", False)
```
Comment (MoritzWillmann) [inline src/squlearn/util/executor.py:1714]:
```suggestion
        use_HQAA = options.get("useHQAA", False)
```

File: tests/util/execution/test_parallel_primitives.py
Comment (MoritzWillmann) [inline tests/util/execution/test_parallel_primitives.py:1]: could be more elaborate

**Review by MoritzWillmann — CHANGES_REQUESTED** [review]
Some code styling changes necessary

File: src/squlearn/qnn/lowlevel_qnn_qiskit.py [hunk 6]
```diff
@@ -309,16 +315,60 @@ def __init__(
...
+                self._primitive = "sampler"
+        else:
+            if primitive.lower() == "estimator":
+                self._estimator = self._executor.get_estimator()
+                self._sampler = None
+                self._primitive = "estimator"
+            elif primitive.lower() == "sampler":
+                self._sampler = self._executor.get_sampler()
+                self._estimator = None
+                self._primitive = "sampler"
```
[the remark this answers is not in the document]
Comment (David-Kreplin) [reply src/squlearn/qnn/lowlevel_qnn_qiskit.py:368]: optree_executor is only a getter from the executor. In previous version, the QNN utilized the primitiv that has been used in the Executor initialization. This is only for backwards compatibility.

File: tests/util/execution/test_parallel_primitives.py
Comment (David-Kreplin) [reply tests/util/execution/test_parallel_primitives.py:1]: Adjusted the testjobs so that the result is also checked.

**Comment by David-Kreplin** [comment]
I hope that is now done

File: src/squlearn/util/executor.py [hunk 7]
```diff
@@ -64,15 +66,17 @@ class Executor:
...
                                                                                                                        In this case the backend has to be provided separately via ``backend=``
                                                                                                                      * A Session, to run the jobs on the Qiskit Runtime service
                                                                                                                      * A Estimator primitive (either simulator or Qiskit Runtime primitive)
                                                                                                                      * A Sampler primitive (either simulator or Qiskit Runtime primitive)
 
-                                                                                                                     Default is the initialization with the :class:`StatevectorSimulator`.
+                                                                                                       Default is the initialization with PennyLane.
```
Comment (MoritzWillmann) [inline src/squlearn/util/executor.py:79]: i would explicitly mention the default qubit device

File: src/squlearn/util/executor.py [hunk 8]
```diff
@@ -88,6 +92,15 @@ class Executor:
...
         seed (Union[int, None]): The seed that is used for finite samples in the execution.
+        qpu_parallelization (Union[int, str, None]): The number of parallel executions on the QPU.
+                                                     If set to ``"auto"``, the number of parallel
+                                                     executions is automatically determined. If set
+                                                     to ``None``, no parallelization is used.
+                                                     Default is ``None``.
+        auto_backend_mode (str): The mode for automatic backend selection. Possible values are:
+
+                                    * ``"quality"``: Automatically selects the best backend for the provided circuit. This is the default value.
+                                    * ``"speed"``: Automatically selects the backend with the smallest queue.
```
Comment (MoritzWillmann) [inline src/squlearn/util/executor.py:103]: fix indentation

File: pyproject.toml
Comment (MoritzWillmann) [inline pyproject.toml:1]: qiskit and numpy dependencies need to be changed again

File: src/squlearn/util/executor.py [hunk 9]
```diff
@@ -410,11 +484,41 @@ def __init__(
...
+            if "ibm" in str(self._backend).lower() or "ibm" in str(self._backend_list).lower():
+                if (
+                    "fake" in str(self._backend).lower()
+                    or "fake" in str(self._backend_list).lower()
+                ):
+                    self._remote = False
+                else:
+                    self._remote = True
             else:
                 self._remote = False
```
Comment (MoritzWillmann) [inline src/squlearn/util/executor.py:496]: needs to be changed to `self._IBMQuantum`, rright?

File: src/squlearn/util/executor.py [hunk 10]
```diff
@@ -410,11 +484,41 @@ def __init__(
 
         # Check if execution is on a remote backend
         if self.quantum_framework == "qiskit":
-            if "ibm" in self.backend_name:
-                self._remote = True
+            if "ibm" in str(self._backend).lower() or "ibm" in str(self._backend_list).lower():
```
[the remark this answers is not in the document]
Comment (MoritzWillmann) [reply src/squlearn/util/executor.py:492]: `backend_name` should remain reserved for the selected backend

File: src/squlearn/util/execution/parallel_estimator.py [hunk 11]
```diff
@@ -0,0 +1,436 @@
+from qiskit.primitives import BaseEstimator
+from qiskit.providers import Options
+from qiskit_ibm_runtime.options import Options as qiskit_ibm_runtime_Options
```
[the remark this answers is not in the document]
Comment (MoritzWillmann) [reply src/squlearn/util/execution/parallel_estimator.py:3]: needs to be changed in all `squlearn`. Refer this to linting issue

File: src/squlearn/util/execution/backend_auto_selection.py [hunk 12]
```diff
@@ -0,0 +1,534 @@
...
+                cost_function=self.cost_function,
+            )
+            self._print(f"{this_score=}")
+            if this_score[0][1] < best_score:
+                best_qc = final_circuit
+                best_backend = backend
+                best_score = this_score[0][1]
+                best_layout = this_layout
+
+        return best_qc, best_backend, best_score, best_layout
```
[the remark this answers is not in the document]
Comment (David-Kreplin) [reply src/squlearn/util/execution/backend_auto_selection.py:499]: Maybe jsut set it None before the loop, or make sure that it is set in the first loop.

File: src/squlearn/util/executor.py [hunk 13]
```diff
@@ -1475,6 +1700,131 @@ def set_seed_for_primitive(self, seed: int = 0):
...
+
+            # duplicate the circuit
+            for _ in range(self._qpu_parallelization - 1):
+                mapped_circuit.tensor(real_circuit, inplace=True)
+
+            info, transpiled_circuit, backend = auto_selection_backend.evaluate(
+                mapped_circuit, mode=mode, use_hqaa=use_hqaa
+            )
+
+            return_circ = circuit
```
Comment (MoritzWillmann) [inline src/squlearn/util/executor.py:1764]: shouldn't this be
```suggestion
            return_circ = transpiled_circuit
```
?

**Review by MoritzWillmann — APPROVED** [review]
Finally 😄

**Pull request merged**